File: src/BackTrack/RemoveInvalidParentheses.java
package BackTrack;

import java.util.*;

// https://leetcode.com/problems/remove-invalid-parentheses/
public class RemoveInvalidParentheses {
    public List<String> removeInvalidParentheses(String s) {
        int left = 0, right = 0, count = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                left++;
            } else if (s.charAt(i) == ')') {
                if (left > 0)
                    left--;
                else
                    right++;
            }
        }
        // Find how many min. no. of changes are required
        count = left + right;
        Set<String> res = new HashSet<>();
        dfs(s, 0, res, new StringBuilder(), count);
        return new ArrayList<>(res);
    }

    boolean isValid(String s) {
        int count = 0;

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(') count++;
            if (c == ')' && count-- == 0) return false;
        }

        return count == 0;
    }

    boolean isValid2(String s) {
        Stack<Character> parenStack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(') parenStack.add(')');
            else if (c == ')') {
                if (parenStack.isEmpty() || parenStack.pop() != c) {
                    return false;
                }
            }
        }
        return parenStack.isEmpty();
    }

    public void dfs(String s, int i, Set<String> res, StringBuilder sb, int count) {
        // draw a decision tree to understand it :)
        if (count < 0) {
            return;
        }
        if (i == s.length()) {
            if (count == 0) {
                if (isValid2(sb.toString())) {
                    res.add(sb.toString());
                }
            }
            return;
        }

        char c = s.charAt(i);
        int len = sb.length();

        if (c == '(' || c == ')') {
            dfs(s, i + 1, res, sb, count - 1);		    // not use ')''(', remove it, so count could be down
        }
        dfs(s, i + 1, res, sb.append(c), count);       // use

        sb.setLength(len);
    }

    public static void main(String[] args) {
        RemoveInvalidParentheses r = new RemoveInvalidParentheses();
        String input = "(a)())()";
        System.out.println(r.removeInvalidParentheses(input));
    }
}

File: py/essential/RemoveInvalidParentheses.py
# https://leetcode.com/problems/remove-invalid-parentheses/description/

from typing import List


class Solution:
    def removeInvalidParentheses1(self, s: str) -> List[str]:
        # leftCount: how many "(" is larger than ")"
        def isValid(input: str) -> tuple[bool, int, int]:
            stack, leftCount, rightCount = [], 0, 0
            for i in range(len(input)):
                if input[i] == "(":
                    stack.append(input[i])
                    leftCount += 1
                if input[i] == ")":
                    if stack and stack[-1] == "(":
                        stack.pop()
                        leftCount -= 1
                    else:
                        rightCount += 1
            return leftCount == rightCount == 0, leftCount, rightCount

        def dfs(input: str, leftCount: int, rightCount: int):
            visited.add(input)
            if isValid(input)[0]:
                res.append(input)
            for i, letter in enumerate(input):
                if letter != "(" and letter != ")":
                    continue
                if (letter == "(" and leftCount <= 0) or (
                    letter == ")" and rightCount <= 0
                ):
                    continue
                if not input[:i] + input[i + 1 :] in visited:
                    dfs(
                        input[:i] + input[i + 1 :],
                        leftCount - (letter == "("),
                        rightCount - (letter == ")"),
                    )

        res = []
        visited = set()
        valid, leftCount, rightCount = isValid(s)

        if valid:
            return [s]
        dfs(s, leftCount, rightCount)
        return res


s = Solution()

print("-----------")
print(s.removeInvalidParentheses1("(a)())()"))
print(s.removeInvalidParentheses1(")))((("))
print(s.removeInvalidParentheses1("(((s)))"))
